cpp/src/hash/hash_constants.hpp

```diff
@@ -60,5 +60,67 @@ __device__ __constant__ md5_hash_constants_type md5_hash_constants[64] = {
   0xf4292244, 0x432aff97, 0xab9423a7, 0xfc93a039, 0x655b59c3, 0x8f0ccc92, 0xffeff47d, 0x85845dd1,
   0x6fa87e4f, 0xfe2ce6e0, 0xa3014314, 0x4e0811a1, 0xf7537e82, 0xbd3af235, 0x2ad7d2bb, 0xeb86d391,
 };
+
+using sha256_word_type = uint32_t;
+
+struct sha256_intermediate_data {
+  uint64_t message_length = 0;
+  uint32_t buffer_length  = 0;
+  uint32_t hash_value[8];
+  uint8_t buffer[64];
+};
+
+__device__ __constant__ sha256_word_type sha256_hash_constants[64] = {
```

Comment:
I'm concerned about using up all of the available constant memory with all of these lookup tables. There's only 64KB available.

Reply:
Also, what will the access pattern be across threads in the same warp? If it is random, rather than uniform, I think just `__device__ const` would be better.